service/history/timerQueueProcessor.go

```diff
@@ -654,17 +654,32 @@ Update_History_Loop:
 				}
 			} else {
 				// See if we have next timer in list to be created.
-				if !td.TaskCreated || td.ActivityID == scheduleID {
+				isHeartBeatTask := timerTask.TimeoutType == int(workflow.TimeoutType_HEARTBEAT)
+
+				// Create next timer task if we don't have one (or)
+				// if current one is HB task and we need to create next HB task for the same.
+				if !td.TaskCreated || (isHeartBeatTask && td.ActivityID == scheduleID) {
```

Comment:
would it be simpler if you simply reset TaskCreated everytime a hearbeat is recorded for an activity ?